utilities/pytorch_model.py

```diff
@@ -200,32 +258,81 @@ def torch_model(self, X_train, X_test, y_train, y_test, params,
+        else:
+            criterion = nn.MSELoss()
         optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
-        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
-                        optimizer, patience=5)
 
         for epoch in range(num_epochs):
             running_loss = 0.0
             for i, (X_batch, y_batch) in enumerate(train_loader):
                 optimizer.zero_grad()
                 y_pred = model(X_batch)
                 loss = criterion(y_pred, y_batch)
+
                 # add l1 loss
                 l1_loss = sum(torch.norm(param, 1)
                                 for name, param in model.named_parameters()
                                 if 'bias' not in name)
                 loss += l1_penalty * l1_loss
+
+                # add L2 network penalty if applicable
+                if self.laplacian is not None:
+                    # get weights w/gradients
+                    w = [param for name, param in model.named_parameters()
+                               if 'bias' not in name][0]
+                    # filter for features that are in the network
+                    w = w[:, self.network_features]
+                    # calculate w^T @ L @ w
+                    network_loss = torch.mm(
+                        w.view(1, -1),
+                        torch.sparse.mm(self.laplacian, w.view(-1, 1)))
+                    loss += (network_penalty * network_loss).view(-1)[0]
+
                 running_loss += loss
                 loss.backward()
                 optimizer.step()
-            scheduler.step(running_loss)
+
+            if learning_curves:
+                # save train loss and test loss on whole dataset after each epoch
+                y_pred_train = model(X_tr)
+                y_pred_test = model(X_ts)
+                self.monitor_['train_loss'].append(float((
+                    criterion(y_pred_train, y_tr)
+                ).detach()))
+                self.monitor_['test_loss'].append(float((
+                    criterion(y_pred_test, y_ts)
+                ).detach()))
+                # also save l1 loss
+                self.monitor_['l1_loss'].append(float((
+                    sum(torch.norm(param, 1)
+                            for name, param in model.named_parameters()
+                            if 'bias' not in name)
+                ).detach()))
+                # also save L2 network loss
+                network_weights = [param for name, param in model.named_parameters()
+                                         if 'bias' not in name][0]
+                network_weights = network_weights[:, self.network_features]
+                network_loss = torch.mm(
+                    w.view(1, -1),
+                    torch.sparse.mm(self.laplacian, w.view(-1, 1)))
+                self.monitor_['network_loss'].append(float((
+                    (network_loss).view(-1)[0]
+                ).detach()))
```

Comment:
You may have already considered this, but you could use a tensorboard [summarywriter](https://pytorch.org/docs/stable/tensorboard.html) to serialize information to a file instead of storing everything in a variable. If you do, you get visualization for free via tensorboard

Reply:
I did consider this, but I was trying to quickly debug a specific issue so I just threw together a simple solution. Longer-term I think using TensorBoard is probably more flexible though - created #18 to switch over in the future.